File: cxsever/www/js/lib/lineardoc/Parser.js
/**
 * @external Contextualizer
 */

import sax from 'sax';
import Builder from './Builder.js';
import { isTransclusion as _isTransclusion, is_inline_empty_tag, isMath, is_reference, isSegment } from './utils.js';

const BLOCK_TAGS = [
	'html', 'head', 'body', 'script',
	// head tags
	// In HTML5+RDFa, link/meta are actually allowed anywhere in the body, and are to be
	// treated as void flow content (like <br> and <img>).
	'title', 'style', 'meta', 'link', 'noscript', 'base',
	// non-visual content
	'audio', 'data', 'datagrid', 'datalist', 'dialog', 'eventsource', 'form',
	'iframe', 'main', 'menu', 'menuitem', 'optgroup', 'option',
	// paragraph
	'div', 'p',
	// tables
	'table', 'tbody', 'thead', 'tfoot', 'caption', 'th', 'tr', 'td',
	// lists
	'ul', 'ol', 'li', 'dl', 'dt', 'dd',
	// HTML5 heading content
	'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup',
	// HTML5 sectioning content
	'article', 'aside', 'body', 'nav', 'section', 'footer', 'header', 'figure',
	'figcaption', 'fieldset', 'details', 'blockquote',
	// other
	'hr', 'button', 'canvas', 'center', 'col', 'colgroup', 'embed',
	'map', 'object', 'pre', 'progress', 'video',
	// non-annotation inline tags
	'img', 'br',
	'wiki-chart'
];

/**
 * Parser to read an HTML stream into a Doc
 *
 * @class
 */
class Parser extends sax.SAXParser {
	/**
	 * @param {Contextualizer} contextualizer Tag contextualizer
	 * @param {Object} options Options
	 */
	constructor(contextualizer, options) {
		super(false, {
			lowercase: true
		});
		this.contextualizer = contextualizer;
		this.options = options || {};
	}

	init() {
		this.rootBuilder = new Builder();
		this.builder = this.rootBuilder;
		// Stack of tags currently open
		this.allTags = [];
	}

	on_open_tag(tag) {
		if (
			// Check if this tag is a child tag of a removable tag
			this.contextualizer.get_context() === 'removable' ||
			// Check if the tag is removable. Note that it is not added to contextualizer yet.
			this.contextualizer.isRemovable(tag)
		) {
			this.allTags.push(tag);
			this.contextualizer.on_open_tag(tag);
			return;
		}

		if (this.options.isolateSegments && isSegment(tag)) {
			this.builder.pushBlockTag({
				name: 'div',
				attributes: {
					class: 'cx-segment-block'
				}
			});
		}

		if (is_reference(tag) || isMath(tag)) {
			// Start a reference: create a child builder, and move into it
			this.builder = this.builder.createChildBuilder(tag);
		} else if (is_inline_empty_tag(tag.name)) {
			this.builder.addInlineContent(
				tag, this.contextualizer.can_segment()
			);
		} else if (this.is_inline_annotation_tag(tag.name, _isTransclusion(tag))) {
			this.builder.pushInlineAnnotationTag(tag);
		} else {
			this.builder.pushBlockTag(tag);
		}

		this.allTags.push(tag);
		this.contextualizer.on_open_tag(tag);
	}

	on_close_tag(tagName) {
		const tag = this.allTags.pop(),
			is_ann = this.is_inline_annotation_tag(tagName, _isTransclusion(tag));

		if (this.contextualizer.isRemovable(tag) || this.contextualizer.get_context() === 'removable') {
			this.contextualizer.on_close_tag(tag);
			return;
		}

		this.contextualizer.on_close_tag(tag);

		if (is_inline_empty_tag(tagName)) {
			return;
		} else if (is_ann && this.builder.inlineAnnotationTags.length > 0) {
			this.builder.popInlineAnnotationTag(tagName);
			if (this.options.isolateSegments && isSegment(tag)) {
				this.builder.pop_block_tag('div');
			}
		} else if (is_ann && this.builder.parent !== null) {
			// In a sub document: should be a span or sup that closes a reference
			if (tagName !== 'span' && tagName !== 'sup') {
				throw new Error('Expected close reference - span or sup tags, got "' + tagName + '"');
			}
			this.builder.finishTextBlock();
			this.builder.parent.addInlineContent(
				this.builder.doc, this.contextualizer.can_segment()
			);
			// Finished with child now. Move back to the parent builder
			this.builder = this.builder.parent;
		} else if (!is_ann) {
			// Block level tag close
			if (tagName === 'p' && this.contextualizer.can_segment()) {
				// Add an empty textchunk before the closing block tag to flush segmentation contexts
				// For example, transclusion based references at the end of paragraphs
				this.builder.addTextChunk('', this.contextualizer.can_segment());
			}
			this.builder.pop_block_tag(tagName);
		} else {
			throw new Error('Unexpected close tag: ' + tagName);
		}
	}

	ontext(text) {
		if (this.contextualizer.get_context() === 'removable') {
			return;
		}
		this.builder.addTextChunk(text, this.contextualizer.can_segment());
	}

	onscript(text) {
		this.builder.addTextChunk(text, this.contextualizer.can_segment());
	}

	/**
	 * Determine whether a tag is an inline annotation or not
	 *
	 * @param {string} tagName Tag name in lowercase
	 * @param {boolean} isTransclusion If the tag is transclusion
	 * @return {boolean} Whether the tag is an inline annotation
	 */
	is_inline_annotation_tag(tagName, isTransclusion) {
		const context = this.contextualizer.get_context();
		// <span> inside a media context acts like a block tag wrapping another block tag <video>
		// See https://www.mediawiki.org/wiki/Specs/HTML/1.7.0#Audio/Video
		if (tagName === 'span' && context === 'media') {
			return false;
		}

		// Audio or Video are block tags. But in a media-inline context they are inline.
		if ((tagName === 'audio' || tagName === 'video') && context === 'media-inline') {
			return true;
		}

		// Styles are usually block tags, but sometimes style tags are used as transclusions
		// Example: T217585. In such cases treat styles as inline to avoid wrong segmentations.
		if (tagName === 'style' && isTransclusion) {
			return true;
		}
		// All tags that are not block tags are inline annotation tags.
		return !BLOCK_TAGS.includes(tagName);
	}
}

export default Parser;

File: cxsever/www/python/lib/lineardoc/parser.py
"""
Parser to read an HTML stream into a Doc.
"""

from lxml import etree
from .builder import Builder
from . import utils as Utils


BLOCK_TAGS = [
    'html', 'head', 'body', 'script',
    # head tags
    'title', 'style', 'meta', 'link', 'noscript', 'base',
    # non-visual content
    'audio', 'data', 'datagrid', 'datalist', 'dialog', 'eventsource', 'form',
    'iframe', 'main', 'menu', 'menuitem', 'optgroup', 'option',
    # paragraph
    'div', 'p',
    # tables
    'table', 'tbody', 'thead', 'tfoot', 'caption', 'th', 'tr', 'td',
    # lists
    'ul', 'ol', 'li', 'dl', 'dt', 'dd',
    # HTML5 heading content
    'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup',
    # HTML5 sectioning content
    'article', 'aside', 'body', 'nav', 'section', 'footer', 'header', 'figure',
    'figcaption', 'fieldset', 'details', 'blockquote',
    # other
    'hr', 'button', 'canvas', 'center', 'col', 'colgroup', 'embed',
    'map', 'object', 'pre', 'progress', 'video',
    # non-annotation inline tags
    'img', 'br'
]


class Parser:
    """Parser to read an HTML stream into a Doc."""

    def __init__(self, contextualizer, options=None):
        """
        Initialize the parser.
        
        Args:
            contextualizer: Tag contextualizer
            options: Options dict
        """
        self.contextualizer = contextualizer
        self.options = options or {}
        self.lowercase = True

    def init(self):
        """Initialize parser state."""
        self.root_builder = Builder()
        self.builder = self.root_builder
        # Stack of tags currently open
        self.all_tags = []

    def write(self, html):
        """
        Parse HTML into the document.
        
        Args:
            html: HTML string to parse
        """
        parser = etree.HTMLParser()
        try:
            tree = etree.fromstring(html.encode('utf-8'), parser)
            self._process_element(tree)
        except Exception:
            # Try with wrapping
            try:
                tree = etree.fromstring(f'<div>{html}</div>'.encode('utf-8'), parser)
                for child in tree:
                    self._process_element(child)
            except Exception as e:
                raise Exception(f'Failed to parse HTML: {e}')

    def _process_element(self, element):
        """Process an element and its children recursively."""
        # Skip comments and other special nodes
        if not isinstance(element.tag, str):
            return
        
        tag_name = element.tag.lower() if self.lowercase else element.tag
        
        # Create tag dict
        tag = {
            'name': tag_name,
            'attributes': dict(element.attrib)
        }
        
        self.on_open_tag(tag)
        
        # Process text content
        if element.text:
            self.on_text(element.text)
        
        # Process children
        for child in element:
            self._process_element(child)
            # Process tail text after child
            if child.tail:
                self.on_text(child.tail)
        
        self.on_close_tag(tag_name)

    def on_open_tag(self, tag):
        """
        Handle open tag event.
        
        Args:
            tag: Tag dict
        """
        if (self.contextualizer.get_context() == 'removable' or
                self.contextualizer.is_removable(tag)):
            self.all_tags.append(tag)
            self.contextualizer.on_open_tag(tag)
            return
        
        if self.options.get('isolateSegments') and Utils.is_segment(tag):
            self.builder.push_block_tag({
                'name': 'div',
                'attributes': {
                    'class': 'cx-segment-block'
                }
            })
        
        if Utils.is_reference(tag) or Utils.is_math(tag):
            # Start a reference: create a child builder, and move into it
            self.builder = self.builder.create_child_builder(tag)
        elif Utils.is_inline_empty_tag(tag['name']):
            self.builder.add_inline_content(
                tag, self.contextualizer.can_segment()
            )
        elif self.is_inline_annotation_tag(tag['name'], Utils.is_transclusion(tag)):
            self.builder.push_inline_annotation_tag(tag)
        else:
            self.builder.push_block_tag(tag)
        
        self.all_tags.append(tag)
        self.contextualizer.on_open_tag(tag)

    def on_close_tag(self, tag_name):
        """
        Handle close tag event.
        
        Args:
            tag_name: Name of tag to close
        """
        if not self.all_tags:
            return
        
        tag = self.all_tags.pop()
        is_ann = self.is_inline_annotation_tag(tag_name, Utils.is_transclusion(tag))
        
        if (self.contextualizer.is_removable(tag) or
                self.contextualizer.get_context() == 'removable'):
            self.contextualizer.on_close_tag(tag)
            return
        
        self.contextualizer.on_close_tag(tag)
        
        if Utils.is_inline_empty_tag(tag_name):
            return
        elif is_ann and len(self.builder.inline_annotation_tags) > 0:
            self.builder.pop_inline_annotation_tag(tag_name)
            if self.options.get('isolateSegments') and Utils.is_segment(tag):
                self.builder.pop_block_tag('div')
        elif is_ann and self.builder.parent is not None:
            # In a sub document: should be a span or sup that closes a reference
            if tag_name not in ('span', 'sup'):
                raise Exception(f'Expected close reference - span or sup tags, got "{tag_name}"')
            self.builder.finish_text_block()
            self.builder.parent.add_inline_content(
                self.builder.doc, self.contextualizer.can_segment()
            )
            # Finished with child now. Move back to the parent builder
            self.builder = self.builder.parent
        elif not is_ann:
            # Block level tag close
            if tag_name == 'p' and self.contextualizer.can_segment():
                # Add an empty textchunk before the closing block tag
                self.builder.add_text_chunk('', self.contextualizer.can_segment())
            self.builder.pop_block_tag(tag_name)
        else:
            raise Exception(f'Unexpected close tag: {tag_name}')

    def on_text(self, text):
        """
        Handle text event.
        
        Args:
            text: Text content
        """
        if self.contextualizer.get_context() == 'removable':
            return
        self.builder.add_text_chunk(text, self.contextualizer.can_segment())

    def on_script(self, text):
        """Handle script text."""
        self.builder.add_text_chunk(text, self.contextualizer.can_segment())

    def is_inline_annotation_tag(self, tag_name, is_transclusion):
        """
        Determine whether a tag is an inline annotation or not.
        
        Args:
            tag_name: Tag name in lowercase
            is_transclusion: If the tag is transclusion
            
        Returns:
            Whether the tag is an inline annotation
        """
        context = self.contextualizer.get_context()
        
        # <span> inside a media context acts like a block tag
        if tag_name == 'span' and context == 'media':
            return False
        
        # Audio or Video are block tags. But in a media-inline context they are inline
        if tag_name in ('audio', 'video') and context == 'media-inline':
            return True
        
        # Styles are usually block tags, but sometimes style tags are used as transclusions
        if tag_name == 'style' and is_transclusion:
            return True
        
        # All tags that are not block tags are inline annotation tags
        return tag_name not in BLOCK_TAGS
